src/nrniv/neuronapi.cpp

```diff
@@ -0,0 +1,466 @@
+#include "../../nrnconf.h"
+#include "nrniv_mf.h"
+#include "nrnmpi.h"
+#include "nrnmpiuse.h"
+#include "ocfunc.h"
+#include "ocjump.h"
+
+#include "parse.hpp"
+
+#include "hocdec.h"
+#include "section.h"
+
+
+// we define these here to allow the API to be independent of internal details
+typedef enum {
+    STACK_IS_STR = 1,
+    STACK_IS_VAR = 2,
+    STACK_IS_NUM = 3,
+    STACK_IS_OBJVAR = 4,
+    STACK_IS_OBJTMP = 5,
+    STACK_IS_USERINT = 6,
+    STACK_IS_SYM = 7,
+    STACK_IS_OBJUNREF = 8,
+    STACK_UNKNOWN = -1
+} stack_types_t;
```

Comment:
Why is this redefined? `neuronapi.h` should be included here, and the issues that that raises should be fixed.

For example, in this file we define something that looks like `nrn_Item* nrn_get_allsec()` but is actually `hoc_Item* nrn_get_allsec()` because of the typedef just below this comment. In `neuronapi.h` we define the same name, `nrn_get_allsec` as being a **pointer to** a function with a different type:
```c
nrn_Item* (*nrn_get_allsec)(void);
```
where here `nrn_Item` is a `struct`, not a `typedef`.

You get away with this for now partly because `extern "C"` stops the (mangled) signature from being used (OK, and because the return type isn't normally mangled and I chose a bad example; but the same applies with e.g. `nrn_new_sectionlist_iterator` that takes an "`nrn_Item*`" argument).

I don't think it's a good idea to try and play these games. The header should declare a public API and the NEURON library should contain/export all the (public) definitions promised by that header, with exactly the promised types.

I understand you're keen to be able to use this in "explicit `dlsym`" mode, which relies on being able to write down the type associated with a particular name and cast to it. I don't know a non-repetitive way of doing this in C, but the mechanical way would be to include a typedef of each API function type next to the function declaration in `neuronapi.h` with some suffix, e.g.
```c
nrn_Item* nrn_get_allsec(void);
typedef nrn_Item* (*nrn_get_allsec_t)(void);
```
I suppose this could be enforced with a `.cpp` file inside NEURON that `static_assert`s that all `foo_t` are the same type as `decltype(&foo)`.

In any case, I think that if code uses `dlsym` to get pointers to these functions, that code should be responsible for owning/managing/organising those pointers, instead of encouraging the use of NEURON-provided global function pointer variables.

Reply:
Also, for the current formulation of declaring function pointer variables in `neuronapi.h`: this will lead to multiple definition errors when one tries to link together multiple object files that included `neuronapi.h`.

See e.g.
- [CPL.2: If you must use C, use the common subset of C and C++, and compile the C code as C++](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rcpl-subset)
- [SF.2: A header file must not contain object definitions or non-inline function definitions](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rs-inline)
- [SF.5: A .cpp file must include the header file(s) that defines its interface](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rs-consistency)

Reply:
Using
```C++
nrn_Item* nrn_get_allsec(void);
typedef nrn_Item* (*nrn_get_allsec_t)(void);
```
should work in Linux and macOS for the "explicit `dlsym`" mode, provided one compiles with `-undefined dynamic_lookup` For Windows, we'd have to do something different, as it would create undefined symbols, but there's probably an optimizer flag to just drop unused symbols.

Reply:
An ugly but easy work-around would be to put all the prototypes behind an `#ifndef` so that they can be blocked if needed.

Reply:
I have minimal experience building stuff like this for Windows, but it would seem surprising if including (and then ignoring/not calling) a declaration of a function caused an undefined symbol error 🤔